pkg/process/metadata/workloadmeta/grpc.go

```diff
@@ -5,12 +5,153 @@
+	for i, proc := range diff.deletion {
+		unsetEvents[i] = &pbgo.ProcessEventUnset{Pid: proc.pid}
+	}
+
+	return setEvents, unsetEvents
+}
+
+// Start starts the GRPCServer to listen for new connections
+func (l *GRPCServer) Start() error {
+	log.Info("Starting Process Entity WorkloadMeta gRPC server")
+	listener, err := getListener(l.config)
+	if err != nil {
+		return err
+	}
+	l.addr = listener.Addr()
+	log.Info("Process Entity WorkloadMeta gRPC server has started listening on", listener.Addr().String())
+
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+
+		err = l.server.Serve(listener)
+		if err != nil {
+			_ = log.Error(err)
+		}
+	}()
+
+	return nil
 }
 
-var _ mockableGrpcListener = (*noopGRPCListener)(nil)
+// Stop stops and cleans up resources allocated by the GRPCServer
+func (l *GRPCServer) Stop() {
+	log.Info("Stopping Process Entity WorkloadMeta gRPC server")
+	l.server.Stop()
+	l.wg.Wait()
+	log.Info("Process Entity WorkloadMeta gRPC server stopped")
+}
+
+// StreamEntities streams Process Entities collected through the WorkloadMetaExtractor
+func (l *GRPCServer) StreamEntities(_ *pbgo.ProcessStreamEntitiesRequest, out pbgo.ProcessEntityStream_StreamEntitiesServer) error {
+	// When connection is created, send a snapshot of all processes detected on the host so far as "SET" events
+	procs, snapshotVersion := l.extractor.GetAllProcessEntities()
+	setEvents := make([]*pbgo.ProcessEventSet, 0, len(procs))
+	for _, proc := range procs {
+		setEvents = append(setEvents, &pbgo.ProcessEventSet{
+			Pid:      proc.pid,
+			Language: &pbgo.Language{Name: string(proc.language.Name)},
+		})
+	}
+
+	syncMessage := &pbgo.ProcessStreamResponse{
+		EventID:   snapshotVersion,
+		SetEvents: setEvents,
+	}
+	err := out.Send(syncMessage)
+	if err != nil {
+		log.Errorf("error sending process entity event: %s", err)
+		return err
+	}
+
+	// Once connection is established, only diffs (process creations/deletions) are sent to the client
+	for {
+		select {
+		case diff := <-l.extractor.ProcessCacheDiff():
+			// Do not send diff if it has the same version of the cache snapshot sent on the connection creation
+			if diff.cacheVersion == snapshotVersion {
+				continue
+			}
 
-type noopGRPCListener struct{}
+			sets, unsets := l.consumeProcessDiff(diff)
+			msg := &pbgo.ProcessStreamResponse{
+				EventID:     diff.cacheVersion,
+				SetEvents:   sets,
+				UnsetEvents: unsets,
+			}
+			err := out.Send(msg)
+			if err != nil {
+				log.Warnf("error sending process entity event: %s", err)
+				return err
+			}
 
-func (l *noopGRPCListener) writeEvents(procsToDelete, procsToAdd []*ProcessEntity) {}
+		case <-out.Context().Done():
+			return nil
+		}
+	}
+}
+
+// getListener returns a listening connection
+func getListener(cfg config.ConfigReader) (net.Listener, error) {
+	address, err := config.GetIPCAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	return net.Listen("tcp", fmt.Sprintf("%v:%v", address, getGRPCStreamPort(cfg)))
```

Comment:
[minor] Let's use the [net.JoinHostPort](https://pkg.go.dev/net#JoinHostPort) function here instead of Sprintf

Reply:
Fixed in https://github.com/DataDog/datadog-agent/pull/17594/commits/c7b47f9e299d2e5faed1e7de18c76499f36bfa78! Thanks for the suggestion!